src/Io/ClientConnectionManager.php

```diff
@@ -0,0 +1,140 @@
+<?php
+
+namespace React\Http\Io;
+
+use Psr\Http\Message\UriInterface;
+use React\EventLoop\LoopInterface;
+use React\EventLoop\TimerInterface;
+use React\Promise\PromiseInterface;
+use React\Socket\ConnectionInterface;
+use React\Socket\ConnectorInterface;
+
+/**
+ * [Internal] Manages outgoing HTTP connections for the HTTP client
+ *
+ * @internal
+ * @final
+ */
+class ClientConnectionManager
+{
+    /** @var ConnectorInterface */
+    private $connector;
+
+    /** @var LoopInterface */
+    private $loop;
+
+    /** @var string[] */
+    private $idleUris = array();
+
+    /** @var ConnectionInterface[] */
+    private $idleConnections = array();
+
+    /** @var TimerInterface[] */
+    private $idleTimers = array();
+
+    /** @var \Closure[] */
+    private $idleStreamHandlers = array();
+
+    /** @var float */
+    private $maximumTimeToKeepAliveIdleConnection = 0.001;
+
+    public function __construct(ConnectorInterface $connector, LoopInterface $loop)
+    {
+        $this->connector = $connector;
+        $this->loop = $loop;
+    }
+
+    /**
+     * @return PromiseInterface<ConnectionInterface>
+     * @throws void
+     */
+    public function connect(UriInterface $uri)
+    {
+        $scheme = $uri->getScheme();
+        if ($scheme !== 'https' && $scheme !== 'http') {
+            return \React\Promise\reject(new \InvalidArgumentException(
+                'Invalid request URL given'
+            ));
+        }
+
+        $port = $uri->getPort();
+        if ($port === null) {
+            $port = $scheme === 'https' ? 443 : 80;
+        }
+        $uri = ($scheme === 'https' ? 'tls://' : '') . $uri->getHost() . ':' . $port;
+
+        // Reuse idle connection for same URI if available
+        foreach ($this->idleConnections as $id => $connection) {
+            if ($this->idleUris[$id] === $uri) {
+                assert($this->idleStreamHandlers[$id] instanceof \Closure);
+                $connection->removeListener('close', $this->idleStreamHandlers[$id]);
+                $connection->removeListener('data', $this->idleStreamHandlers[$id]);
+                $connection->removeListener('error', $this->idleStreamHandlers[$id]);
+
+                assert($this->idleTimers[$id] instanceof TimerInterface);
+                $this->loop->cancelTimer($this->idleTimers[$id]);
+                unset($this->idleUris[$id], $this->idleConnections[$id], $this->idleTimers[$id], $this->idleStreamHandlers[$id]);
+
+                return \React\Promise\resolve($connection);
+            }
+        }
+
+        // Create new connection if no idle connection to same URI is available
+        return $this->connector->connect($uri);
```

Comment:
Just thinking out loud here, but isn't this faster?

```suggestion
        if (array_key_exists($uri, $this->idleUris) {
            $connection = $this->idleConnections[$this->idleUris[$uri]];
            
            assert($this->idleStreamHandlers[$id] instanceof \Closure);
            $connection->removeListener('close', $this->idleStreamHandlers[$id]);
            $connection->removeListener('data', $this->idleStreamHandlers[$id]);
            $connection->removeListener('error', $this->idleStreamHandlers[$id]);
            assert($this->idleTimers[$id] instanceof TimerInterface);
            $this->loop->cancelTimer($this->idleTimers[$id]);
            unset($this->idleUris[$id], $this->idleConnections[$id], $this->idleTimers[$id], $this->idleStreamHandlers[$id]);

            return \React\Promise\resolve($connection);
        }

        // Create new connection if no idle connection to same URI is available
        return $this->connector->connect($uri);
```

Reply:
It's an tempting thought, but the `ClientConnectionManager` may end up multiple connections to the same host if there are concurrent connections, hence the `$uri` can't be used as a unique key. I'm sure there's some room for future improvements here, but figured it makes sense to get this initial version out of the door first.

Reply:
Good point, and it might only work with HTTP/2. However I still see room for improvement, does it make sense to have a list per URI so we don't have to iterate over the entire list? The more connections we might have opened the slower this will become.

It is also something we can fix later, it's not a blocker for me.

Reply:
Fair enough, just amended my changeset to remove this so we can focus on the core functionality here instead, this is something we can still discuss in the future when introducing PHPStan (#472). 👍

Reply:
Agree that improving this in the future might be worthwhile. HTTP/2 would likely also start with multiple connections (see e.g. https://daniel.haxx.se/blog/2016/08/18/http2-connection-coalescing/), but let's take a look at this again once we prepare a PR for #99.

My vote would be to get this in as is and file any future improvements as separate PRs once we have more input 👍

Reply:
> It is also something we can fix later, it's not a blocker for me.

> My vote would be to get this in as is and file any future improvements as separate PRs once we have more input 👍

GMTA 😉